### DISCUSSION TITLE
Modify note or access bestiary entry to track resources

### DISCUSSION THREAD

**Discussion by ThaddaeusWiedemer** [Obsidian-TTRPG-Community/fantasy-statblocks#418] [Q&A] [answered]
I'm looking for a way to track the resources (spell slots, items, abilities with charges) of a monster (for the sake of simplicity, assume a unique monster like a named NPC) through the _rendered_ statblock.

For example, I can render a checkbox for an ability with 2 charges using a JavaScript block. When clicked, the rendered checkbox updates the `monster` property, for example, a `monster.resource` variable. Currently, I'm not sure whether there's any way to make this variable persistent between renders of the statblock (e.g., because the note was closed and reopened, or because the statblock was edited).

I can think of two ways to achieve this in principle:

1. **Make persistent in note**. The checkbox could have an EventListener that modifies the note's content to update the statblock YAML itself with the status of the resource. This would require somehow getting a handle on the current note from the JavaScript block.

2. **Make persistent in bestiary**. I saw in [this discussion](https://github.com/javalent/fantasy-statblocks/discussions/115#discussioncomment-8259336) that the `monster` property in the JavaScript block is a pointer to the Bestiary object. So a programmatically added `monster.resource` should, in principle, be saved and could be available for the next render. Still, at the moment it seems that by the time a JavaScript block is executed, `monster` has already been overwritten again.

Is either way achievable with the plugin's current capabilities or with an additional plugin like Dataview? Does anyone have pointers on how this could be done?

**Comment by valentine195** [comment, marked as answer]
A rendered stay block will listen for changes to the creature in the bestiary and re-render. You could use something like metabind to update a creature defined in frontmatter. The plugin will re-parse the frontmatter within the next 2 seconds after and then update the view.

**Reply by ThaddaeusWiedemer** [reply]
Thanks for the hint! Based on this, I was able to cook up the following code which works for inline statblocks; maybe somebody else finds it useful as well:

```javascript
monster.uses = monster.uses ?? {};

monster._getFile = async function () {
  const creature = await window.FantasyStatblocks.getCreature(monster.name);
  return file = app.vault.getFileByPath(creature.path);
}

monster._updateUses = function (name, val) {
  monster.uses[name] = val;

  console.log(monster.uses);

  let usesText = '```statblock\nuses:\n';
  for (const [key, val] of Object.entries(monster.uses)) {
    usesText += `  ${key}: ${val}\n`;
  }

  const pattern = /(```statblock\n(?:uses:\n(?:  [^:]+: [^\n]+\n)*)?)/;

  monster._getFile().then(f => app.vault.process(f, (data) => {
    return data.replace(pattern, usesText);
  }));
}
```

The basic idea is that I add a `uses` object to the monster that will be serialized as, e.g., 

```
uses:
  Breath Weapon: 0
  Spell L4: 4
```

Whenever this object is updated, the function finds the path to the monster's file (via the monster's name through the Fantasy Statblocks API), gets the file handle through `app.vault.getFileByPath()`, and modifies the statblock-codeblock to include the serialized uses.

Since the underlying file is modified directly, this updates the resources no matter where the stat block is rendered, so it also works in the creature view of an encounter, for example.

The only slightly annoying thing is that the entire statblock is re-rendered as soon as the file is modified, so changing a resource will re-render the entire statblock
